File: code/java/ch04/practice/Practice03Temperature.java
package ch04.practice;

import java.util.Scanner;

/**
 * Practice 03: Temperature Conversion
 * ==============================
 * Chapter 4: Functions
 *
 * PROBLEM
 * -------
 * Write a temperature converter that handles conversions between
 * Celsius (C), Fahrenheit (F), and Kelvin (K).
 *
 * Use helper functions cToF(), fToC(), cToK(), kToC() to build
 * all six conversion paths.
 *
 * Formulas:
 *   C to F:  F = C * 9/5 + 32
 *   F to C:  C = (F - 32) * 5/9
 *   C to K:  K = C + 273.15
 *   K to C:  C = K - 273.15
 *
 * INPUT FORMAT
 * ------------
 * One line: value fromUnit toUnit
 * Example: 100 C F
 *
 * OUTPUT FORMAT
 * -------------
 * Print the converted value rounded to 2 decimal places.
 * Print -1.0 for invalid units or same-unit conversion.
 *
 * CONSTRAINTS
 * -----------
 * - value is a floating-point number
 * - fromUnit and toUnit are one of "C", "F", "K"
 *
 * EXAMPLES
 * --------
 * Input:  100 C F
 * Output: 212.00
 *
 * Input:  32 F C
 * Output: 0.00
 *
 * Input:  0 C K
 * Output: 273.15
 *
 * INSTRUCTIONS
 * ------------
 * 1. Implement cToF(), fToC(), cToK(), kToC() helpers.
 * 2. Build all 6 conversions by chaining helpers (e.g., F to K = fToC then cToK).
 * 3. Return -1.0 for invalid or same-unit conversion.
 * The main method handles input/output -- don't change it.
 */
public class Practice03Temperature {

    /** Convert Celsius to Fahrenheit. */
    public static double cToF(double c) {
        // TODO: Implement
        return 0.0;
    }

    /** Convert Fahrenheit to Celsius. */
    public static double fToC(double f) {
        // TODO: Implement
        return 0.0;
    }

    /** Convert Celsius to Kelvin. */
    public static double cToK(double c) {
        // TODO: Implement
        return 0.0;
    }

    /** Convert Kelvin to Celsius. */
    public static double kToC(double k) {
        // TODO: Implement
        return 0.0;
    }

    /**
     * Convert a temperature value between units.
     *
     * @param value    the temperature value
     * @param fromUnit the source unit ("C", "F", or "K")
     * @param toUnit   the target unit ("C", "F", or "K")
     * @return the converted value, or -1.0 for invalid/same-unit
     */
    public static double solve(double value, String fromUnit, String toUnit) {
        // TODO: Replace this with your solution
        return -1.0;
    }

    // ── Do not change anything below this line ──────────────────────
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        double value = sc.nextDouble();
        String fromUnit = sc.next();
        String toUnit = sc.next();
        double result = solve(value, fromUnit, toUnit);
        System.out.printf("%.2f%n", result);
        sc.close();
    }
}

File: code/cpp/ch04/practice/practice_03_temperature.cpp
/*
 * Practice 3: Temperature Converter
 * ==================================
 * Chapter 4: Functions
 *
 * PROBLEM:
 *   Convert between Celsius (C), Fahrenheit (F), and Kelvin (K).
 *   Write helper functions: c_to_f, f_to_c, c_to_k, k_to_c.
 *   Use helpers to build the general solve() converter.
 *   Round result to 1 decimal place.
 *
 * FORMULAS:
 *   C to F: F = C * 9/5 + 32
 *   F to C: C = (F - 32) * 5/9
 *   C to K: K = C + 273.15
 *   K to C: C = K - 273.15
 *
 * EXAMPLES:
 *   solve(100.0, "C", "F")  -> 212.0
 *   solve(32.0, "F", "C")   -> 0.0
 *   solve(0.0, "C", "K")    -> 273.2   (273.15 rounds to 273.2)
 *   solve(300.0, "K", "F")  -> 80.3
 *   solve(50.0, "C", "C")   -> 50.0    (same unit)
 *   solve(50.0, "X", "C")   -> -1.0    (invalid unit)
 *
 * CONSTRAINTS:
 *   - Valid units: "C", "F", "K"
 *   - Return -1.0 for invalid units
 *   - Round to 1 decimal place using: round(result * 10.0) / 10.0
 */

#include <cmath>
#include <iostream>
#include <string>
using namespace std;

// TODO: Write helpers: c_to_f, f_to_c, c_to_k, k_to_c

/**
 * Converts temperature from one unit to another.
 * Returns -1.0 for invalid units.
 */
double solve(double value, string from_unit, string to_unit) {
    // TODO: Replace this with your solution
    return -1.0;
}

// -- Do not change anything below this line --------------------------
int main() {
    double value;
    string from_unit, to_unit;
    cin >> value >> from_unit >> to_unit;
    double result = solve(value, from_unit, to_unit);
    cout << result << endl;
    return 0;
}
